create-plugin: Add provisioning scaffold (#529)
grafana/plugin-tools: docusaurus/docs/publish-a-plugin/provide-provisioning.md

---
id: provide-provisioning
title: Provide a test environment
description: How to add provisioning to your plugin to speed up your plugin review process.
keywords:
  - github
  - provisioning
  - provision
  - review
sidebar_position: 5
---

# Plugin provisioning

Developers often ask us how long it takes for a plugin to be reviewed for publishing to the Grafana [plugin catalog](https://grafana.com/plugins). Although we [can't provide estimates](https://grafana.com/developers/plugin-tools/publish-a-plugin/publish-a-plugin#how-long-does-it-take-to-review-a-submission), we are always looking for ways to reduce cycle times.

By far the most time consuming aspect of a plugin's review is the creation of a suitable test environment so we can verify its behavior. This step often involves a number of back-and-forth conversations between the plugin developer and the review team.

> To drastically improve the review time, developers can provide this themselves by [_provisioning_](https://grafana.com/docs/grafana/latest/administration/provisioning/#provision-grafana) their plugin. Provisioning refers to the process of preparing and configuring a test environment within the plugin's [Docker development environment]((https://grafana.com/developers/plugin-tools/get-started/set-up-development-environment).

Reviewer: ```suggestion
To drastically improve the review time, developers can provide this themselves by [_provisioning_](https://grafana.com/docs/grafana/latest/administration/provisioning/#provision-grafana) their plugin. Provisioning refers to the process of preparing and configuring a test environment within the plugin's [Docker development environment](https://grafana.com/developers/plugin-tools/get-started/set-up-development-environment).
```

---
id: provide-provisioning

> title: Provision your plugin

Reviewer: I'm not sure about the focus being on "provision(ing)" - the mechanism vs "Provide a test environment" - the intent.

Reviewer 2: If Marcus is okay with it, I'm good with "Provide a test environment".

> description: How to add provisioning to your plugin to potentially speed up your plugin review process.

Reviewer: ```suggestion
description: How to add provisioning to your plugin to speed up your plugin review process.
```

keywords:
  - github
  - provisioning
  - provision
  - review
sidebar_position: 5
---

# Plugin provisioning

> Developers of community plugins often ask us these questions:

Reviewer: This section comes across as unnecessarily negative to me and that we're just shifting the pain . 

Suggestion - (note the lazy links since I'm doing this in the web editor), may also need editing for more conversational tone as well.

>Developers often ask us how long it takes for a plugin to be reviewed for publishing to the Grafana `[plugin catalog](https://grafana.com/plugins)`. Whilst we `[can't provide estimates](link to the faq)` - we are always looking for ways to reduce cycle times, such as through additional automation.
>
>By far the most time consuming aspect of a plugin's review is the creation of a suitable test environment so we can verify its behaviour, often involving a number of back-and-forth conversations between the developer and the review team. To drastically improve the time it takes for a plugin to be reviewed, developers can provide this themselves through including the appropriate resources and configuration to be `[provisioned](https://grafana.com/docs/grafana/latest/administration/provisioning/#provision-grafana)` within the `[docker development environment](link to docker dev env)` we generate with the `create-plugin` tool.

Reviewer 2: nice. edited a bit.

- How long does it take to review plugins for inclusion in the Grafana [plugin catalog](https://grafana.com/plugins)?
- How can Grafana Labs speed up the time it takes us to review their submissions?

We have heard you. Our response to these questions has been to allow plugin developers themselves to speed the review process by creating a test environment.

The mechanism for including such an environment with your plugin submission is called _provisioning_. Provisioning refers to the process of preparing and configuring your plugin environment to meet specific requirements, making it easier for reviewers to assess your plugin comprehensively.

> ## Mechanism to provide for provisioning

Reviewer: I struggle to read this title, and again wonder why we need to have the focus on provisioning

```suggestion
## Mechanism to provide a test environment
```

Grafana can be configured to have resources installed and enabled through a mechanism known as provisioning, where resources are configured in a YAML file under a `/provisioning` directory.

> We have done this for some time in the [create-plugin tool](../get-started/get-started.mdx) for app plugins so that they can be automatically enabled (that is, provisioned) when using our Docker development environment.

Reviewer: The reference to past/ "we're now" makes this read more like a blog/point-in-time article. Suggest removing and just saying describing what the tool does by default and the new command. Optionally can include a version number such as "Since `create-plugin` v2.7.0, we generate..."

Reviewer 2: done.

We’re now extending provisioning to cover all plugin types (apps, scenes-apps, datasources, panels), and to include a sample dashboard as part of create-plugin.

> ## Why should plugin developers care about this?

Reviewer: I'd move this above the mechansim

Reviewer 2: done.

There are several benefits to provisioning:

- **Much faster review times.** If you provision your plugin prior to submission, your wait for the review will be much shorter.
- **Easier contributions.** By having an out-of-the-box working example available, would-be contributors to your plugin can easily experiment with additions to the plugin and raise pull requests.
- **Easier set up for e2e tests.** By provisioning dashboards, e2e tests can run against specific scenarios across local development and in CI.

> - **Enhanced transparency.** We have found that provisioned plugins make it easier for tech-savvy users to understand how the plugin works.

Reviewer: ```suggestion
- **Improved usability.** We have found that provisioned plugins make it easier for tech-savvy users to understand how the plugin works.
```

_maybe_ more appropriate than transparency

Reviewer 2: (breaks out thesaurus) I like "Improved clarity" best.

## What do plugin developers need to do?

:::note

Provisioning is not required; it's an optional part of the plugin submission process that will speed the review process.

:::

1. When you run the create-plugin tool, it will generate a folder called `provisioning` with additional files based on the plugin type selected.
1. When you run the Docker development environment, these files are used to automatically install (and if applicable, _enable_) your plugin and a sample dashboard.

> 1. We recommended that you use and update the sample dashboard to continuously verify behavior as part of your development process. And, as appropriate, use the dashboard to configure your plugin so that it can return data.

Reviewer: ```suggestion
1. We recommended that you use and update the sample dashboard to continuously verify behavior as part of your development process. And, as appropriate, configure your plugin so that it can return data.
```

Reviewer 2: done.

> 1. We recommend that you submit the provisioning files along with your plugin for review.

Reviewer: How?

Author: I have deleted this last point, because we will have all the files including provisioning via the source code link

:::note

> In the case where a plugin has been scaffolded with a previous version of create-plugin, a new command can be run to retrospectively add the provisioning files.

Reviewer: Couple questions:
1. What happens if the user already has a `/provisioning` folder? 
2. Does `update` do anything with content under `/provisioning`

Author: If we have already have /provisioning folder, updating will abort with the message "You plugin already has provisioning files located under ..." - this will be a case only for "old" plugins which were created before those changes. It was made just for backwards compatibilities

Reviewer: Sorry I'm not sure I understand, are you saying that Running `npx @grafana/create-plugin@latest update` will result in an error if the user has ` /provisioning` folder?

Author: Oh, sorry, no.` npx @grafana/create-plugin@latest update` does not interfere with provisioning at all. But if you run `npx @grafana/create-plugin@latest provisioning` it will check if provisioning is already there

Reviewer: Perfect - just wanted to be sure 😄
